Use reusable GitHub Actions for testing

ekohl opened the pull request:
This reduces the duplication between projects. For example, this ensures testing on Ruby 2.7 which previously didn't happen.

There are various preparation PRs, see the commits for details.

Right now this is incomplete because it loses the caching step that happens on a push to master. It also still uses the repository in my own namespace. This is just to see if it works.

nofaralfasi commented:
Upon reviewing the PR, it seems that it has been open for quite some time without recent activity. In order to keep our repository organized and up-to-date, we've decided to close this draft PR for now. 
Feel free to reopen it if you decide to continue this work.

ekohl commented on .github/workflows/ruby_tests.yml:
```diff
@@ -1,81 +1,20 @@
-        run: |
-          bundle exec rake test:foreman_ansible -v --trace
-          bundle exec rake test TEST="test/unit/foreman/access_permissions_test.rb" -v --trace
-      - name: 'Upload logs'
-        uses: actions/upload-artifact@v2
-        if: failure()
-        with:
-          name: logs
-          path: log/*.log
-          retention-days: 5
```
This is not yet part of the reusable action.

ekohl replied:
By now it is.

ekohl commented on .github/workflows/ruby_tests.yml:
```diff
@@ -1,81 +1,20 @@
-        uses: actions/setup-node@v1
-        with:
-          node-version:  ${{ matrix.node-version }}
-      - name: Install npm dependencies
-        run: npm install
-      - name: Install plugin's npm dependencies
-        run: npm install --omit=dev
-        working-directory: ${{ github.workspace }}/foreman_ansible
-      - name: Prepare test env
-        run: |
-          bundle exec rake db:create
-          bundle exec rake db:migrate
-          bundle exec rake db:test:prepare
-      - name: Prepare webpack
-        run: bundle exec rake webpack:compile
```
This is not yet part of the reusable action

ekohl replied:
By now we do have webpack compilation.

ekohl reviewed (commented):
Still a TODO, but this needs to happen. The current action is using wrong versions for NodeJS today. It will get worse once we add Ruby 3 to the test matrix.

ekohl commented:
Rebased. I've split off https://github.com/theforeman/foreman_ansible/pull/679 & https://github.com/theforeman/foreman_ansible/pull/680 to separate PRs so they can be merged already.

ekohl commented on lib/tasks/foreman_ansible_tasks.rake:
```diff
@@ -3,10 +3,11 @@
+    test_dir = File.join(__dir__, '..', '..', 'test')
     t.libs << ['test', test_dir]
     t.pattern = "#{test_dir}/**/*_test.rb"
+    t.test_files = [Rails.root.join('test/unit/foreman/access_permissions_test.rb')]
```
There's a related discussion in https://github.com/theforeman/actions/pull/16 on the best way to do this.

ekohl commented on lib/tasks/foreman_ansible_tasks.rake:
```diff
@@ -3,10 +3,11 @@
   desc 'Foreman Ansible plugin tests'
-  Rake::TestTask.new(:foreman_ansible) do |t|
-    test_dir = File.join(File.dirname(__FILE__), '../..', 'test')
+  Rake::TestTask.new(:foreman_ansible => 'db:test:prepare') do |t|
```
I wonder why this dependency is needed, given other plugins don't do it. I think we can simply drop it instead.

evgeni commented on .github/workflows/ruby_tests.yml:
```diff
@@ -1,81 +1,20 @@
-  DATABASE_URL: postgresql://postgres:@localhost/test
-  DATABASE_CLEANER_ALLOW_REMOTE_DATABASE_URL: true
+
+on: pull_request
```
```suggestion
on:
  pull_request:
  push:
    branches:
      - 'master'
      - '*-stable'
```

otherwise there won't be no cache, which I find rather important to have :)

ekohl replied:
We should expand the actions readme with this (and multible Foreman versions)

evgeni replied:
Are you spying on my vim buffers?!

evgeni replied:
https://github.com/theforeman/actions/pull/34

evgeni commented on .github/workflows/ruby_tests.yml:
```diff
@@ -1,81 +1,20 @@
-    strategy:
-      fail-fast: false
-      matrix:
-        foreman-core-branch: [develop]
```
the old workflow had a way to run with multiple Foreman versions (even if the list was set to `develop` only at the time of this PR), should we resurect this?

@nofaralfasi what do you think?

ekohl replied:
Unless we actually utilize the matrix, I'd leave it for the future.

nofaralfasi commented:
Can we merge this PR?
